File: 424. Longest Repeating Character Replacement/solution.py
"""
You are given a string s and an integer k. You can choose any character of the string and change it to any other uppercase English character. 

You can perform this operation at most k times.

Return the length of the longest substring containing the same letter you can get after performing the above operations.
"""

from collections import defaultdict


class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        """
        Time: O(N)
        Space: O(1)
        """
        map = defaultdict(int)
        left = 0
        max_len = 0
        max_count = 0
        for right in range(len(s)):
            map[s[right]] += 1
            max_count = max(max_count, map[s[right]])
            if right - left + 1 - max_count > k:
                map[s[left]] -= 1
                left += 1
            max_len = max(max_len, right-left+1)
        return max_len

File: 424. Longest Repeating Character Replacement/solution.go
package main

func max(a, b int)int {
	if a > b {
		return a
	} else {
		return b
	}
}

func characterReplacement(s string, k int) int {
    hMap := make(map[string]int)
	var maxCount int
	var maxLen int
	var left int = 0
	var str = []rune(s)

	for right := 0; right < len(str); right++ {
		hMap[string(str[right])] += 1
		maxCount = max(maxCount, hMap[string(str[right])])
		if right - left + 1 - maxCount > k {
			hMap[string(str[left])] -= 1
			left++
		}
		 maxLen = max(maxLen, right - left + 1)
	}
	return maxLen
}